File: ios/ObjC+UIKit/IntegrationDemo_ObjC_UIKit/ViewController.m
//
//  ViewController.m
//  IntegrationDemo_ObjC_UIKit
//
//  Created by Jeroen Verbeek on 2/22/23.
//

#import "ViewController.h"
#import "PlacementCell.h"
#import <TapResearchSDK/TapResearchSDK.h>
#import "NativeWallViewController.h"

@interface ViewController () <UITextFieldDelegate,
                              UITableViewDelegate,
                              UITableViewDataSource,
                              TapResearchContentDelegate
                              >

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UITextField *textField;
@property (weak, nonatomic) IBOutlet UILabel *placementStatus;

@property (strong, nonatomic) NSMutableArray *knownPlacements;
@property NSString *surveysPlacement;

@end

@implementation ViewController

- (void)viewDidLoad {
	[super viewDidLoad];

	self.knownPlacements = @[
		@"default-placement",
		@"interstitial-placement",
		@"banner-placemenet",
		@"floating-interstitial-placement"
	].mutableCopy;
	self.surveysPlacement = @"earn-center";
	[NSNotificationCenter.defaultCenter addObserver:self selector:@selector(handleRefresh:) name:@"refresh" object:nil];
	self.textField.placeholder = @"Placement Tag";
	self.textField.delegate = self;
}

- (void)viewWillAppear:(BOOL)animated {
	[super viewWillAppear:animated];

	UIBarButtonItem *button = [[UIBarButtonItem alloc] initWithTitle:@"Surveys?" style:UIBarButtonItemStylePlain target:self action:@selector(refresh)];
	[self.navigationItem setRightBarButtonItem:button];
	[self.tableView reloadData];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {

	if ([segue.identifier isEqualToString:@"ShowSurveys"]) {
		NativeWallViewController *vc = ((NativeWallViewController*)segue.destinationViewController);
		vc.placementTag = self.surveysPlacement;
	}
}

//MARK: - UITextFieldDelegate

- (void)textFieldDidChangeSelection:(UITextField *)textField {

	if (self.placementStatus.text && self.placementStatus.text.length > 0) {
		self.placementStatus.text = nil;
	}
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {

	[self showPlacement];
	return YES;
}

//MARK: - Actions and button handlers

-(void)refresh {
	[self.tableView reloadData];
}

- (IBAction)showPlacement {

	NSString *placementTag = self.textField.text;

	if (placementTag && placementTag.length > 0) {
		if ([TapResearch canShowContentForPlacement:placementTag error:^(NSError * _Nullable error) {
			// Handle error, this is an optional error block
		}]) {
			[TapResearch showContentForPlacement:placementTag delegate:self completion:^(NSError * _Nullable error) {
				if (error) {
					self.placementStatus.text = [NSString stringWithFormat:@"%ld, %@", (long)error.code, error.localizedDescription];
				}
				else {
					if (![self.knownPlacements containsObject:placementTag]) {
						[self.knownPlacements addObject:placementTag];
						dispatch_async( dispatch_get_main_queue(), ^{
							[self.tableView reloadData];
						});
					}
					self.placementStatus.text = nil;
				}
			}];
		}
		else {
			self.placementStatus.text = @"No content for placement";
		}
	}
}

//MARK: - Tableview delegate and datasource

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {

	if (indexPath.section == 0) {
		return [PlacementCell cellForTableView:tableView placementTag:self.knownPlacements[indexPath.row] andInfo:nil];
	}
	else {
		return [PlacementCell cellForTableView:tableView placementTag:self.surveysPlacement andInfo:nil];
	}
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {

	[tableView deselectRowAtIndexPath:indexPath animated:YES];

	if (indexPath.section == 0) {
		NSString *placementTag = self.knownPlacements[indexPath.row];
		if ([TapResearch canShowContentForPlacement: placementTag error:^(NSError * _Nullable error) {
			// Handle error, this is an optional error block
		}]) {
			[TapResearch showContentForPlacement:placementTag delegate:self customParameters:@{@"custom_param_1" : @"test text", @"custom_param_3" : @12} completion:^(NSError * _Nullable error) {
				if (error) {
					NSLog(@"Error on showContent: %ld, %@", (long)error.code, error.localizedDescription);
				}
			}];
		}
		else {
			NSLog(@"Placement not ready");
		}
	}
	else if (indexPath.section == 1) {
		[self performSegueWithIdentifier:@"ShowSurveys" sender:self.surveysPlacement];
	}
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {

	if (section == 0) {
		return @"Known Placements";
	}
	return @"Surveys";
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {

	BOOL hasSurveys = [TapResearch hasSurveysFor:self.surveysPlacement errorHandler:^(NSError * _Nullable error) { }];
	if (hasSurveys) {
		return 2;
	}
	return 1;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {

	if (section == 0) {
		return self.knownPlacements.count;
	}
	return 1;
}

//MARK: - TapResearchContentDelegate

- (void)onTapResearchContentDismissedForPlacement:(NSString * _Nonnull)placement {
	NSLog(@"onTapResearchContentDismissedForPlacement(%@)", placement);
}

- (void)onTapResearchContentShownForPlacement:(NSString * _Nonnull)placement {
	NSLog(@"onTapResearchContentShownForPlacement(%@)", placement);
}

- (void)handleRefresh:(NSNotification*)n {
	[self.tableView reloadData];
}

@end

File: ios/FullFeatureSet/Swift_UIKit_Notif/IntegrationDemo_Swift_UIKit_Notif/ViewController.swift
//
//  ViewController.swift
//  IntegrationDemo_Swift_UIKit_Notif
//
//  Created by Jeroen Verbeek on 2/22/23.
//

import UIKit
import TapResearchSDK

class ViewController : UIViewController,
					   UITextFieldDelegate,
					   UITableViewDelegate,
					   UITableViewDataSource,
					   TapResearchGrantBoostResponseDelegate,
					   LogPrint
{

	@IBOutlet weak var tableView: UITableView!
	@IBOutlet weak var placementTextField: UITextField!
	@IBOutlet weak var placementStatus: UILabel!
	@IBOutlet weak var boostTextField: UITextField!
	@IBOutlet weak var boostStatus: UILabel!

	var surveysPlacement: String = "earn-center"
	let showSurveysSegue: String = "ShowSurveys"
	var knownPlacements: [String] = [
		"earn-center",
		"default-placement",
		"interstitial-placement",
		"banner-placement",
		"floating-interstitial-placement"
	]

	override func viewDidLoad() {
		super.viewDidLoad()

		placementTextField.placeholder = "Placement Tag"
		placementTextField.delegate = self
		boostTextField.placeholder = "Boost Tag"
		boostTextField.delegate = self
	}

	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)
		
		NotificationCenter.default.addObserver(self, selector: #selector(self.onTapResearchContentShown(_:)), name: TapResearch.TapResearchContentShown, object: nil)
		NotificationCenter.default.addObserver(self, selector: #selector(self.onTapResearchContentDismissed(_:)), name: TapResearch.TapResearchContentDismissed, object: nil)

		navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Surveys?", style: .plain, target: self, action: #selector(refresh))
	}

	override func viewDidDisappear(_ animated: Bool) {
		super.viewDidDisappear(animated)

		NotificationCenter.default.removeObserver(self, name: TapResearch.TapResearchContentShown, object: nil)
		NotificationCenter.default.removeObserver(self, name: TapResearch.TapResearchContentDismissed, object: nil)
	}

	override func prepare(for segue: UIStoryboardSegue, sender: Any?) {

		if segue.identifier == showSurveysSegue {
			if let vc: NativeWallViewController = segue.destination as? NativeWallViewController {
				vc.placementTag = surveysPlacement
			}
		}
	}
	
	//MARK: - UITextFieldDelegate

	func textFieldDidChangeSelection(_ textField: UITextField) {

		switch textField {
			case boostTextField:
				if let _ = boostStatus.text {
					boostStatus.text = nil
				}
			case placementTextField:
				if let _ = placementStatus.text {
					placementStatus.text = nil
				}
			default:
				return
		}
	}

	func textFieldShouldReturn(_ textField: UITextField) -> Bool {

		switch textField {
			case boostTextField:
				grantBoost()
				return true
			case placementTextField:
				showPlacement()
				return true
			default:
				return false
		}
	}

	//MARK: - Actions and button handlers

	@objc func refresh() {
		tableView.reloadData()
	}

	@IBAction func grantBoost() {
		guard let text: String = boostTextField.text else { return }

		TapResearch.grantBoost(text, delegate: self) { (error: NSError?) in
			if let error {
				self.boostStatus.text = "\(error.code) \(error.localizedDescription)"
			}
		}
	}

	@IBAction func showPlacement() {
		guard let text: String = placementTextField.text else { return }

		let canShow: Bool = TapResearch.canShowContent(forPlacement: text) { (error: NSError?) in
			// Handle error, this is an optional error block, if there is an error false is returned by function.
		}
		if canShow {
			//let customParameters = ["param1": 123, "param2": "abc"] as [String : Any]
			TapResearch.showContent(forPlacement: text/*, customParameters: customParameters*/) { (error: NSError?) in
				if let error = error {
					self.placementStatus.text = "\(error.code) \(error.localizedDescription)"
				}
				else {
					if !(self.knownPlacements as NSArray).contains(text) {
						self.knownPlacements.append(text)
						DispatchQueue.main.async(execute: { () -> Void in
							self.tableView.reloadData()
						})
					}
				}
			}
		}
		else {
			placementStatus.text = "No content for placement"
		}
	}

	//MARK: - Table delegate and datasource
	
	func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		
		self.tableView.deselectRow(at: indexPath, animated: true)

		if indexPath.section == 1 {
			performSegue(withIdentifier: showSurveysSegue, sender: surveysPlacement)
			return
		}

		if TapResearch.canShowContent(forPlacement: knownPlacements[indexPath.row]) {
			TapResearch.showContent(forPlacement: knownPlacements[indexPath.row], customParameters: ["custom_param_1" : "test text", "custom_param_3" : 12]) { (error: NSError?) in
                if let error = error {
					self.logPrint("Error on showContent: \(error.code) \(error.localizedDescription)")
                }
			}
		}
		else {
			logPrint("Placement not ready")
		}
	}

	func numberOfSections(in tableView: UITableView) -> Int {

		if TapResearch.hasSurveys(for: surveysPlacement) {
			return 2
		}
		return 1
	}

	func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {

		if section == 1 {
			return 1
		}
		return knownPlacements.count
	}

	func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {

		if section == 1 {
			return "Surveys"
		}
		return "Placements"
	}
	
	func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

		if indexPath.section == 1 {
			return PlacementCell.cell(tableView: tableView, indexPath: indexPath, placement: surveysPlacement)
		}
		return PlacementCell.cell(tableView: tableView, indexPath: indexPath, placement: knownPlacements[indexPath.row])
	}

	//MARK: - TapResearchContentDelegate

	@objc func onTapResearchContentShown(_ notification: Notification) {
		if let placementTag: String = notification.userInfo?[TapResearch.PlacementTagKey] as? String {
			logPrint("placementTag = \(placementTag)")
		}
	}

	@objc func onTapResearchContentDismissed(_ notification: Notification) {
		if let placementTag: String = notification.userInfo?[TapResearch.PlacementTagKey] as? String {
			logPrint("placementTag = \(placementTag)")
		}
	}

	//MARK: - TapResearchGrantBoostResponseDelegate

	func onTapResearchGrantBoostResponse(_ response: TapResearchSDK.TRGrantBoostResponse) {

		if response.success {
			self.boostStatus.text = "\(response.boostTag): success!"
		}
		else {
			if let error = response.error {
				self.boostStatus.text = "\(response.boostTag): \(error.localizedDescription)"
			}
			else {
				self.boostStatus.text = "\(response.boostTag): unkown error"
			}
		}
	}

}
